offchain processing macro and example for sov-nft-module

dubbelosix opened the pull request:
# Description
This PR contains the code for `offchain` macro which enables creating functions that can update and manage offchain state/computation that doesn't impact on-chain state or consensus. 
Decorating a function with the offchain macro causes it to no-op unless the "offchain" feature is explicitly enabled. When the offchain feature is enabled while starting the rollup as described in `offchain_readme.md`, the code is executed.

Components added
* offchain macro in sov-module-macros
* offchain.rs inside the sov-nft-module for updating postgres tables in sync with transactions
* init_db.sql which has the scripts to create the necessary tables for offchain processing
* utils/nft-utils - a binary that can be used to submit transactions to the sequencer
* simple_client exists with `send_transaction`, which sends a single transaction in a batch. added `send_transactions` which takes a vector of transactions, serializes them into a batch and submits to the sequencer. optionally, send_transactions can also take a parameter that controls the batch size. For example `Some(3)` would mean that multiple batches of up to 3 transactions each are submitted, and `None` would mean that there is micro batching, and all transactions are submitted.

Outstanding Issues
* There is a timing/concurrency issue with the sequencer and the way it reads from state to accept/reject transactions, so the nft-utils script contains a sleep of 500ms between sends
* sequencer also performs an execute so offchain processing is being done twice. This is not an issue for any of the tables except "top_owners" since the operation is not idempotent (increments are done twice, once in batch_builder and once more in sov_modules_stf_template. investigating options to avoid this double processing

## Testing
The script added to create collections and mint NFTs at `utils/nft-utils` can also be used to test

## Docs
Docmentation specific to using and testing the offchain macro is at `module-system/module-implementations/sov-nft-module/offchain_readme.md`

Commit:
BlockScout integration fixes (#917)

* Fixes

* retesteth config

Signed-off-by: Filippo Costa <filippo@sovlabs.io>

---------

Signed-off-by: Filippo Costa <filippo@sovlabs.io>

bkolad committed:
modules_api: simplify public key definition (#918)

* simplify publi key definition

* private key

bkolad committed:
`EVM`: Add missing docs (#914)

* rename pending_block to block_env

* make evm internals private

* refactor query.rs

* AccountData doc

* query.rs docs

* fix exports

* add missing docs

* cargo fmt

* Add readme

* Update README.md

* Update README.md

* better docs

preston-evans98 committed:
Cleanup prover docs/logs (#919)

* Cleanup prever docs/logs

* fmt

dependabot[bot] committed:
Bump markdown from 1.0.0-alpha.13 to 1.0.0-alpha.14 (#924)

Bumps [markdown](https://github.com/wooorm/markdown-rs) from 1.0.0-alpha.13 to 1.0.0-alpha.14.
- [Release notes](https://github.com/wooorm/markdown-rs/releases)
- [Commits](https://github.com/wooorm/markdown-rs/compare/1.0.0-alpha.13...1.0.0-alpha.14)

---
updated-dependencies:
- dependency-name: markdown
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump rayon from 1.7.0 to 1.8.0 (#923)

Bumps [rayon](https://github.com/rayon-rs/rayon) from 1.7.0 to 1.8.0.
- [Changelog](https://github.com/rayon-rs/rayon/blob/master/RELEASES.md)
- [Commits](https://github.com/rayon-rs/rayon/compare/rayon-core-v1.7.0...rayon-core-v1.8.0)

---
updated-dependencies:
- dependency-name: rayon
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump prost-build from 0.11.9 to 0.12.1 (#930)

Bumps [prost-build](https://github.com/tokio-rs/prost) from 0.11.9 to 0.12.1.
- [Release notes](https://github.com/tokio-rs/prost/releases)
- [Commits](https://github.com/tokio-rs/prost/compare/v0.11.9...v0.12.1)

---
updated-dependencies:
- dependency-name: prost-build
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump parking_lot from 0.11.2 to 0.12.1 (#927)

Bumps [parking_lot](https://github.com/Amanieu/parking_lot) from 0.11.2 to 0.12.1.
- [Changelog](https://github.com/Amanieu/parking_lot/blob/master/CHANGELOG.md)
- [Commits](https://github.com/Amanieu/parking_lot/compare/0.11.2...0.12.1)

---
updated-dependencies:
- dependency-name: parking_lot
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

vlopes11 committed:
feat: use json as manifest constants instead of toml (#922)

* feat: use json as manifest constants instead of toml

Currently, we parse the manifest file as TOML. However, we will use JSON
for the genesis format. For improved consistency, it is desirable to
have the constants manifest file with the same format.

* fix CI lints

* use parent as argument for error reporting

* refactor gas config parse to return declaration

* fix ci lints

dependabot[bot] committed:
Bump tungstenite from 0.20.0 to 0.20.1 (#931)

Bumps [tungstenite](https://github.com/snapview/tungstenite-rs) from 0.20.0 to 0.20.1.
- [Changelog](https://github.com/snapview/tungstenite-rs/blob/master/CHANGELOG.md)
- [Commits](https://github.com/snapview/tungstenite-rs/compare/v0.20.0...v0.20.1)

---
updated-dependencies:
- dependency-name: tungstenite
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

orkunkilic committed:
EVM: Implement eth_call (#921)

* implement eth_call

* implement tests and error handling for eth_call

* move errors inside evm crate

* cleanup result

* fix call env

* comment test differences

* rebase

* Revert "rebase"

This reverts commit 44e41b23dac2a68839cad896ef6c46c20ca410cc.

* fix unused imports

* fix review notes

* use mix_hash as prevrandao

* improve test s

dubbelosix committed:
offchain processing

dubbelosix committed:
cargo lock

preston-evans98 commented on full-node/sov-sequencer/src/utils.rs:
```diff
@@ -38,6 +38,39 @@ impl SimpleClient {
+    pub async fn send_transactions<Tx: BorshSerialize>(
+        &self,
+        txs: Vec<Tx>,
+        micro_batch: Option<usize>,
```
nit: The name `micro_batch` doesn't convey that this is the chunk size very clearly. Maybe it would be clearer to call it `chunk_size`? 

We could also look at changing the API to `fn send_batches(&self, batches: Vec<Vec<Tx>>` and then we wouldn't need the extra param and logic here

preston-evans98 replied:
Update: We never use `micro_batch`. Can we just remove it?

dubbelosix replied:
I use the second parameter primarily while switching between MockDA and celestia.
switched the logic to use Vec<Vec<Tx>> so that the chunking can be handled by the caller

preston-evans98 commented on module-system/module-implementations/sov-nft-module/Cargo.toml:
```diff
@@ -21,6 +21,9 @@ sov-modules-api = { path = "../../sov-modules-api" }
 sov-state = { path = "../../sov-state" }
 
+postgres = {version = "0.19.7", optional = true}
+tokio = {version = "1.32.0", features=["full"],optional = true}
```
nit: formatting
```
postgres = { version = "0.19.7", optional = true }
tokio = { version = "1.32.0", features=["full"], optional = true }
```

preston-evans98 commented on module-system/module-implementations/sov-nft-module/offchain_readme.md:
```diff
@@ -0,0 +1,56 @@
+## Offchain testing
```
This readme could use some context on what the user is supposed to be accomplishing by following these steps. Can we add an intro to orient the user?

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/call.rs:
```diff
@@ -87,6 +88,8 @@ impl<C: Context> NonFungibleToken<C> {
         )?;
         self.collections
             .set(&collection_address, &collection, working_set);
+        std::thread::sleep(std::time::Duration::from_millis(300));
```
Two comments here: 
1. If we're going to introduce a sleep into code, we should definitely add a comment explaining it! 
2. I don't think we can safely add a sleep into the critical path of processing a call message. I don't want to block this tutorial on something unrelated to its functionality (adding synchronization to the sequencer), but I don't feel good about this. Users running the NFT demo will see 3 tps, and it won't be obvious why

dubbelosix replied:
oh yeah, this was just to illustrate the issue but it got checked in accidentally. removed now!

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/init_db.sql:
```diff
@@ -0,0 +1,47 @@
+-- Drop existing tables if they exist
```
nit: The indentation of create table is all over the place. Is that fixable, or does some linter do this automatically?

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+use crate::utils::get_collection_address;
+use crate::{Collection, CollectionAddress, Nft};
+
+// CREATE TABLE collection (
+//     collection_address TEXT PRIMARY KEY,
+//     collection_name TEXT NOT NULL,
+//     creator_address TEXT NOT NULL,
+//     frozen BOOLEAN NOT NULL,
+//     metadata_url TEXT,
+//     supply BIGINT NOT NULL
+// );
```
nit: Let's remove commented code or add some explanation of what it's doing here.

preston-evans98 replied:
nit: This comment is a little unclear. Does it go with the `track_collection function?` If so, why isn't it a doc comment. Also, let's add some explanation explaining what it means - is this a description of the logic that the track collection function implements?

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+
+#[offchain]
+pub fn track_collection<C: sov_modules_api::Context>(collection: &Collection<C>) {
+    // data extraction
```
nit: `// data extraction` -> `// Extract the necessary metadata from the collection`

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+// );
+
+#[offchain]
+pub fn track_collection<C: sov_modules_api::Context>(collection: &Collection<C>) {
```
nit: `track_collection` implies something ongoing. How about `put_collection` or `insert_collection`?

preston-evans98 replied:
Also, this function is missing a doc comment

dubbelosix replied:
changed the name to update_collection!

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+    })
+}
+
+// CREATE TABLE nft (
+//     collection_address TEXT NOT NULL,
+//     nft_id BIGINT NOT NULL,
+//     metadata_url TEXT,
+//     owner TEXT NOT NULL,
+//     frozen BOOLEAN NOT NULL,
+//     PRIMARY KEY (collection_address, nft_id)
+// );
```
nit: comment is unclear

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+// );
+
+#[offchain]
+pub fn track_nft<C: sov_modules_api::Context>(nft: &Nft<C>) {
```
nit: Same comment as above about funciton name (maybe use "insert" or "put"?). Also, missing doc comment

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
```
nit: Can we `tracing::error!` instead of `print`ing in all this cases, or do we want to avoid a `tracing` dependency here?

dubbelosix replied:
dependency should be fine! the more offchain or convenience libs and methods we can use, the more it shows the benefits of using the offchain component (in fact they can even add sentry logging etc and would not need to rely on emitting events or using on-chain compute). 
will add tracing in

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+pub fn track_nft<C: sov_modules_api::Context>(nft: &Nft<C>) {
+    let collection_address = nft.get_collection_address().to_string();
+    let nft_id = nft.get_token_id();
+    let owner = nft.get_owner().to_string();
+    let frozen = nft.is_frozen();
+    let metadata_url = nft.get_token_uri();
+    tokio::task::block_in_place(|| {
+        if let Ok(conn_string) = std::env::var("POSTGRES_CONNECTION_STRING") {
+            match postgres::Client::connect(&conn_string, NoTls) {
+                Ok(mut client) => {
+                    let result = client.execute(
+                        "INSERT INTO nfts (\
+                        collection_address, nft_id, metadata_url,\
+                        owner, frozen)\
+                        VALUES ($1, $2, $3, $4, $5)\
+                        ON CONFLICT (collection_address, nft_id)\
+                        DO UPDATE SET metadata_url = EXCLUDED.metadata_url,\
+                                      owner = EXCLUDED.owner,\
+                                      frozen = EXCLUDED.frozen",
+                        &[
+                            &collection_address,
+                            &(nft_id as i64),
+                            &metadata_url,
+                            &owner,
+                            &frozen,
+                        ],
+                    );
+                    if let Err(e) = result {
+                        println!("Failed to execute query: {}", e);
+                    }
+                }
+                Err(e) => {
+                    println!("Failed to connect to the database: {}", e);
+                }
+            }
+        } else {
+            println!("Environment variable POSTGRES_CONNECTION_STRING is not set");
+        }
+    })
+}
```
Both of these `track_` functions are long and mostly duplicated. I think we can break them down and then factor some code into a shared helper. 

It looks to me like the two components are: 
1. Build a sql query from the data (depends on the shape of the item we're tracking)
2. Try to execute that query (can take query as a simple argument, identical in both cases)

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+    })
+}
+
+// CREATE TABLE top_owners (
```
nit: Same issue with comment

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+// );
+
+#[offchain]
+pub fn update_top_owners<C: sov_modules_api::Context>(
```
nit: missing doc comment

preston-evans98 commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,168 @@
+#[offchain]
+pub fn update_top_owners<C: sov_modules_api::Context>(
+    collection_address: &CollectionAddress<C>,
+    owner_count_incr: Option<&[(String, u64)]>,
+    owner_count_decr: Option<&[(String, u64)]>,
```
Question: why do we use two vecs instead of using a signed integer?

dubbelosix replied:
ah yeah, this logic is changed now with merging update_nft and update_top_owners (original reason was to decerement counts of owner1 and increment counts of owner2 for a transfer, but that logic doesnt matter right now)

preston-evans98 commented on module-system/sov-modules-macros/src/lib.rs:
```diff
@@ -263,3 +265,9 @@ pub fn address_type(_attr: TokenStream, item: TokenStream) -> TokenStream {
 }
+
+#[proc_macro_attribute]
+pub fn offchain(_attr: TokenStream, item: TokenStream) -> TokenStream {
```
nit: Missing doc comment

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+use sov_rollup_interface::mocks::MockDaSpec;
+use sov_sequencer::utils::SimpleClient;
+
+const COLLECTION_1: &str = "Sovereign Squirrel Syndicate";
+const COLLECTION_2: &str = "Celestial Dolphins";
+const COLLECTION_3: &str = "Risky Rhinos";
```
nice :)

preston-evans98 commented on utils/nft-utils/Cargo.toml:
```diff
@@ -0,0 +1,31 @@
+[package]
```
We have a lot of packages in the workspace already. Did we look at options that don't add this at the top level? Maybe we can make this a binary in the other nft package?

dubbelosix replied:
had a discussion with team and it would be good to come up with a more generalized structure to place util libs and util bins for each module. created an issue to track this https://github.com/Sovereign-Labs/sovereign-sdk/issues/957 for now, resolving the comment and keeping the structure at `sovereign/utils/nft-utils`

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+const COLLECTION_2: &str = "Celestial Dolphins";
+const COLLECTION_3: &str = "Risky Rhinos";
+
+const DUMMY_URL: &str = "http://foobar.storage";
+
+const PK1: [u8; 32] = [
+    199, 23, 116, 41, 227, 173, 69, 178, 7, 24, 164, 151, 88, 149, 52, 187, 102, 167, 163, 248, 38,
+    86, 207, 66, 87, 81, 56, 66, 211, 150, 208, 155,
+];
+const PK2: [u8; 32] = [
+    92, 136, 187, 3, 235, 27, 9, 215, 232, 93, 24, 78, 85, 255, 234, 60, 152, 21, 139, 246, 151,
+    129, 152, 227, 231, 204, 38, 84, 159, 129, 71, 143,
+];
+const PK3: [u8; 32] = [
+    233, 139, 68, 72, 169, 252, 229, 117, 72, 144, 47, 191, 13, 42, 32, 107, 190, 52, 102, 210,
+    161, 208, 245, 116, 93, 84, 37, 87, 171, 44, 30, 239,
+];
+
+fn get_collection_metadata_url(collection_address: &str) -> String {
+    format!("{}/collection/{}", DUMMY_URL, collection_address)
+}
+
+fn get_nft_metadata_url(collection_address: &str, nft_id: u64) -> String {
+    format!("{}/nft/{}/{}", DUMMY_URL, collection_address, nft_id)
+}
```
This package mixes real functionality and mocks. Can we separate it cleanly into module with real helpers ("create collection", etc.), and then a binary with tge dummy data and main function? That way, someone could depend on the helpers and get a working implementation by just passing some actual data instead of mocks

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+    format!("{}/nft/{}/{}", DUMMY_URL, collection_address, nft_id)
+}
+
+async fn create_collection(
```
nit: missing doc comment

preston-evans98 replied:
I don't think this function needs to be async. Am I missing something?

preston-evans98 replied:
Also, this name is a bit misleading. Can we rename to something which conveys that this is just generating a call message, not actually creating the collection `build_create_collection_msg` or something?

dubbelosix replied:
ah yeah, the asyncs were part of a prior version where create_collection also ended up sending the transaction directly using simple_client. originally tested with celestia docker and it was too slow, so i added the logic to fetch the transactions and send batches of multiple. both the name and the async ended up remaining. fixed now!

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+    )
+}
+
+async fn mint_nft(
```
nit: missing doc comment

preston-evans98 replied:
Also, Same question about async as above

preston-evans98 replied:
Same comment about naming as above

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+    )
+}
+
+async fn transfer_nft(
```
nit: missing doc comment

preston-evans98 replied:
Same question about async as above

preston-evans98 replied:
Same comment about naming as above

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+    transactions.push(create_collection(&creator_pk, nonce, COLLECTION_3).await);
+
+    client.send_transactions(transactions, None).await.unwrap();
+    // sleep is necessary because of how the sequencer currently works
+    // without the sleep, there is a concurrency issue and some transactions would be ignored
+    thread::sleep(Duration::from_millis(1000));
```
This comment is great. Can we add an issue to track the problem and link it here as well?

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+        creator_pk.default_address().as_ref(),
+    );
+
+    #[allow(clippy::explicit_counter_loop)]
```
Can we use the loop counter instead of using separate `owner_1_nonce` and `nft_id` variables here? Will make clippy happy and remove a little code.

preston-evans98 commented on utils/nft-utils/src/main.rs:
```diff
@@ -0,0 +1,217 @@
+    thread::sleep(Duration::from_millis(1000));
+
+    let mut transactions = vec![];
+    for _ in 0..15 {
+        nonce += 1;
+        transactions.push(
+            mint_nft(
+                &creator_pk,
+                nonce,
+                COLLECTION_1,
+                nft_id,
+                &owner_1_pk.default_address(),
+            )
+            .await,
+        );
+        nft_id += 1;
+    }
```
nit: We have several of these loops with mutable nonces etc. it's not the end of the world, but feels a bit messy. Maybe we can simplify with an iterator, and/or make a helper function?

bkolad committed:
Run tests with prover enabled.  (#932)

* Remove demo-rollup dep from demo-prover

* add methods to demo-rollup

* fix cargo hack

* fix cargo hack

* Cargo.toml

* install toolchain

* install toolchain

* modify workflow

* update cargo.toml

* rust.yaml

* Add guest-mock

* use real prover in bank tests

* native

* fix lint

* update yaml

* fic coverage

* ELF in CI

* update bench get_guest_options()

* update build.rs

* remove local from demo-rollup

* fix lint

* remove ide_setup

* Remove riscv32im-risc0-zkvm-elf

vlopes11 committed:
fix: require `native` when applicable for all targets (#937)

* fix: require `native` when applicable for all targets

Prior to this commit, some checks of all-targets without the feature
`native` would break.

This commit introduces a fix for every workspace member to be consistent
with the feature set.

* change feature requirement to self-dev-dep

* update deps

* fix dupl celestia dependencies

dubbelosix committed:
some fixes

dubbelosix committed:
move common code into lib

bkolad committed:
Remove default trait bound from Module (#941)

bkolad commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,142 @@
+/// Syncs an NFT to the corresponding table "nfts" in postgres
+/// Additionally, this function also has logic to track the counts of NFTs held by each user
+/// in each collection.
+#[offchain]
```
Can this macro be used with `Module` attributes. I am thinking that the `AccesorryXX` should be always behind offchain. @neysofu wdt?

neysofu replied:
I very much agree, I think accessory state should be zero-cost and only written to during native execution. The current model works but we need to integrate it more tightly with `offchain`. Tracked in https://github.com/Sovereign-Labs/sovereign-sdk/issues/944.

bkolad replied:
Cool, @dubbelosix feel free to resolve this comment. We will do it is as #944.

bkolad commented on module-system/module-implementations/sov-nft-module/src/offchain.rs:
```diff
@@ -0,0 +1,142 @@
+
+            // Update NFT information after handling top_owners logic
+            let _ = client.execute(
+                "INSERT INTO nfts (\
```
What about extracting all the sql strings as constants in a separate `sql.rs` file?

bkolad committed:
Remove demo-rollup-local job from CI (#943)

dubbelosix committed:
added issue to the nft script

dubbelosix committed:
add nft cli binary to nft-utils

citizen-stig committed:
Delta's go first, remove `get` from StateCheckpoint (#953)

preston-evans98 committed:
Make genesis config serializable (#956)

* Make genesis config serializable

* Evm config serde

* Fix chain state integ test

* fix vec-setter

bkolad committed:
Introduce: `PublicKeyHex` in `sov-modules-api` (#954)

* Add PublicKeyHex

* PubKeyHex impl

* Update DefaultPublicKey::from_str

* Add tests

* Remove println

* Add doc

dubbelosix committed:
Revert "add nft cli binary to nft-utils"

This reverts commit d443009db9b438112d388e4e78e19d2dfc19fe3f.

orkunkilic committed:
estimate-gas signed (#947)

dubbelosix committed:
move nft-utils back to sovereign/nft-utils and move sql to sql.rs

dubbelosix committed:
remove un-necessary deps in nft-utils

orkunkilic committed:
EVM: Implement account related RPC (#958)

* implement account endpoints

* test account endpoints

* remove unnecessary conversions

vlopes11 committed:
feature: gas meter (#795)

* feat: add gas meter to working set

This commit introduces `GasMeter`, encapsulated by `WorkingSet`.

It will allow the user to consume scalar gas from the working set, and
define arbitrary price parsed from a constants.json manifest file at
compilation. At each compilation, the `ModuleInfo` derive macro will
parse such file, and set the gas price configuration.

* fix lint fmt

* fix ci test expected error string

* update default context to 2 dimensions

dubbelosix committed:
remove include and make it a module

dubbelosix committed:
fix packages.yml

bkolad committed:
Read `accounts` genesis from a file. (#959)

* PrivateKeyHex in accounts

* Add test_config_serialization

* Add accounts.json

* fix ci

* fix CI

* cleanup

* Arbitrary PublicKeyHex

* PublicKeyHex impl

dubbelosix committed:
fix a wrong merge conflict

dubbelosix committed:
Merge branch 'nightly' into dub/offchain_processing
